wallet: Prefer full destination groups in coin selection

fjahr opened the pull request:
Fixes #17603 (together with #17843)

In the case of destination groups of >10 outputs existing in a wallet with `avoid_reuse` enabled, the grouping algorithm is adding left-over outputs as an "incomplete" group to the list of groups even when a full group has already been added. This leads to the strange behavior that if there are >10 outputs for a destination the transaction spending from that will effectively use `len(outputs) % 10` as inputs for that transaction. 

From the original PR and the code comment I understand the correct behavior should be the usage of 10 outputs. I opted for minimal changes in the current code although there maybe optimizations possible for cases with >20 outputs on a destination this sounds like too much of an edge case right now.

fjahr committed:
wallet: Prefer full destination groups in coin selection

When a wallet uses avoid_reuse and has a large number of outputs in
a single destination, it groups these outputs in OutputGroups that
are no larger than OUTPUT_GROUP_MAX_ENTRIES. The goal is to spend
as many outputs as possible from the destination while not breaking
consensus due to a huge number of inputs and also not surprise the
use with high fees. If there are n outputs in a destination and
n > OUTPUT_GROUP_MAX_ENTRIES then this results in one or many groups
of size OUTPUT_GROUP_MAX_ENTRIES and possibly one group of size
< OUTPUT_GROUP_MAX_ENTRIES.

Prior to this commit the coin selection in the case where
n > OUTPUT_GROUP_MAX_ENTRIES was skewed towards the one group of
size < OUTPUT_GROUP_MAX_ENTRIES if it exists and the amount to be
spent by the transaction is smaller than the aggregate of those
of the group size < OUTPUT_GROUP_MAX_ENTRIES. The reason is that
the coin selection decides between the different groups based on
fees and mostly the smaller group will cause smaller fees.

The behavior that users of the avoid_reuse flag seek is that the
full groups of size OUTPUT_GROUP_MAX_ENTRIES get used first. This
commit implements this by pretending that the small group has
a large number of ancestors (one smallet than the maximum allowed
for this wallet). This dumps the small group to the bottom of the
list of priorities in the coin selection algorithm.

instagibbs commented on test/functional/wallet_avoidreuse.py:
```diff
@@ -240,5 +242,26 @@ def test_fund_send_fund_send(self):
 
+    def test_destination_groupings(self):
+        '''
+        If a destination has more than 10 outputs, exactly 10 outputs should
```
what assumptions does this test have?

Seems like it assumes the wallet has no more than 10*21 BTC? In other words, those sends completely drain the wallet of other UTXO addresses. This should be asserted and/or programmatically done.

kallewoof replied:
It took me awhile to realize what is being stated here.

If a destination has more than 10 outputs, the 10-output groups should always be picked before the <10-output group, if any such exists.

I.e. if a destination has 19 outputs, it will result in one 10-output group and one 9-output group. The second should never ever be used unless the former is also used.

fjahr replied:
I have expanded the test and the comments and hope it's more clear now.

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -3977,6 +3977,7 @@ std::vector<OutputGroup> CWallet::GroupOutputs(const std::vector<COutput>& outpu
     std::vector<OutputGroup> groups;
     std::map<CTxDestination, OutputGroup> gmap;
     CTxDestination dst;
+    std::set<CTxDestination> full_groups;
```
ISTM a cleaner approach would be to move `groups` down before the return and just keep everything in the map. When there are more than 10 outputs in a group, just stop adding to the group. This would also handle the case of 20 outputs.

Unless our goal is to have multiple groups when more than 10 outputs for a destination?

cc @kallewoof

kallewoof replied:
The intention was that if there were a ton of spends to the same address, these would be chunked into randomized 10 UTXO chunks, to avoid creating too large transactions. So yeah, the intention was to have multiple groups for the same destination in cases where there were 11+ UTXOs for it.

kallewoof commented:
The reported issue seems like things are not working as they should for sure, but I don't see how this code fixes it. Ultimately, if I want to send 10.5 bitcoin to you, and I have 500 UTXO:s of  1 BTC each all sending to A, the grouping part should result in 50 groups of 10 random UTXO:s each (each group worth 10 BTC), and two of these should be selected when sending, and ALL of the outputs should be marked as used afterwards.

Your code seems to change this behavior to only result in one 10 BTC group per destination, which means the above send will result in "insufficient funds".

kallewoof commented:
I investigated this further. See https://github.com/bitcoin/bitcoin/issues/17603#issuecomment-569867141

fjahr commented:
> I investigated this further. See [#17603 (comment)](https://github.com/bitcoin/bitcoin/issues/17603#issuecomment-569867141)

Thanks, that's what I found as well, I should have described it in greater detail.

> The reported issue seems like things are not working as they should for sure, but I don't see how this code fixes it. Ultimately, if I want to send 10.5 bitcoin to you, and I have 500 UTXO:s of 1 BTC each all sending to A, the grouping part should result in 50 groups of 10 random UTXO:s each (each group worth 10 BTC), and two of these should be selected when sending, and ALL of the outputs should be marked as used afterwards.
> 
> Your code seems to change this behavior to only result in one 10 BTC group per destination, which means the above send will result in "insufficient funds".

My code does handle this case correctly. You describe the behavior that would occur if I had implemented it as @achow101 suggests. The "full groups" (10 outputs) are already pushed into `groups` [earlier](https://github.com/bitcoin/bitcoin/pull/17824/files#diff-b2bb174788c7409b671c46ccc86034bdR3992) and only the "incomplete" group of <10 outputs would be added [later](https://github.com/bitcoin/bitcoin/pull/17824/files#diff-b2bb174788c7409b671c46ccc86034bdR4006). So we are collecting all the full groups but skip the last one if it is not full.

However, your comment made me realize another shortcoming of the current code: If there are 11 outputs of 1 BTC and we want to send 10.5 BTC then we see the insufficient funds error because we are "forgetting" the incomplete group. To fix this I would suggest implementing the following: the last, incomplete group, is pushed into one of the full groups, resulting in one of the groups being up to 19 outputs large. "Worst case" scenario that I see: we want to send a very small amount from the destination and we have exactly 19 outputs, then we would use all the 19 outputs. I think this should be ok? I will wait for your feedback before pushing the code for it.

See adapted tests I ran with the current code for the examples discussed above: https://gist.github.com/fjahr/d56b68e58b4275ef52cd5f4eb50b1d4e

kallewoof commented:
I'm not sure this is the right approach here. It seems like marking the output groups as 'partial' when there are >10 outputs would address this. A simple approach would be to explicitly think of <max groups for multi-group destinations as "0-conf". The current code would then actively try to avoid using it except as a last resort.

kallewoof commented:
I would do something like this
```diff
diff --git a/src/wallet/wallet.cpp b/src/wallet/wallet.cpp
index 3954f6626..3c88ff9c0 100644
--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -3976,6 +3976,7 @@ bool CWalletTx::IsImmatureCoinBase() const
 std::vector<OutputGroup> CWallet::GroupOutputs(const std::vector<COutput>& outputs, bool single_coin) const {
     std::vector<OutputGroup> groups;
     std::map<CTxDestination, OutputGroup> gmap;
+    std::set<CTxDestination> gfull;
     CTxDestination dst;
     for (const auto& output : outputs) {
         if (output.fSpendable) {
@@ -3990,6 +3991,7 @@ std::vector<OutputGroup> CWallet::GroupOutputs(const std::vector<COutput>& outpu
                 if (gmap[dst].m_outputs.size() >= OUTPUT_GROUP_MAX_ENTRIES) {
                     groups.push_back(gmap[dst]);
                     gmap.erase(dst);
+                    gfull.insert(dst);
                 }
                 gmap[dst].Insert(input_coin, output.nDepth, output.tx->IsFromMe(ISMINE_ALL), ancestors, descendants);
             } else {
@@ -3998,7 +4000,16 @@ std::vector<OutputGroup> CWallet::GroupOutputs(const std::vector<COutput>& outpu
         }
     }
     if (!single_coin) {
-        for (const auto& it : gmap) groups.push_back(it.second);
+        for (const auto& it : gmap) {
+            auto& group = it.second;
+            if (gfull.count(it.first) > 0 && group.m_outputs.size() < OUTPUT_GROUP_MAX_ENTRIES) {
+                // make this unattractive as we want coin selection to avoid it if possible
+                group.m_depth = 0;
+                group.m_from_me = false;
+                group.m_ancestors = 9999;
+            }
+            groups.push_back(group);
+        }
     }
     return groups;
 }
```

fjahr reviewed (commented):
> Edit: I didn't test, but I assume your example with "not working" above is for your patch, not the current master.

Yeah, that's right.

> I would do something like this

That is a good idea, I did not think of that possibility before. But I still see a problem there: if there are flags activated like `-walletrejectlongchains` or `-txconfirmtarget` then we will have the same issue as with my code described above. The incomplete group will be ignored and the user may get an insufficient funds error when there are actually enough funds available. And since `DEFAULT_TX_CONFIRM_TARGET` is set to 6 I think this would happen for every wallet that has not set this value to 0. If only using `group.m_ancestors = 9999;` at least this will work for the default wallet configuration. I am not sure how widely `-walletrejectlongchains` is used.

Honestly, I am not sure yet which solution I prefer. The upsides of my described approach:
1. No matter what the configuration of the wallet is, there will not be a false "Insufficient funds" error.
2. I think it does a better job of using funds from the same address together. 

An example to explain 2.: 
- A wallet has 11 outputs of 1 BTC each at `reused_address`
- It also has 1 output of 1 BTC at `not_reused_address`
- The user wants to send 10.5 BTC and uses `avoid_reuse`
- With the `m_ancestors` approach, the wallet would send 10 BTC from `reused_address` and 1 BTC from `not_reused_address`, indicating these addresses could be controlled by the same user
- With my suggestion, there would be 11 BTC group only from `reused_address` selected for the inputs

This stands against the downsides of my approach: 
1. At worst we are using `2 * max - 1` inputs when `max` could have been enough.
2. It is a more complicated change than yours.

Let me know if I am missing something. I am not sure how much weight each of these arguments carry. I think the perfect solution would only be possible by making a distinction between <max groups and max groups in coin selection and that seems like overkill given the impact of the issue. 

I am pushing an updated version implementing your solution with only `m_ancestors` and and an improved test and hope to get some more feedback. But I am happy to make further changes.

kallewoof commented:
You're probably right that it would not end up using the final group even if it could. I think that's okay, in this case. As for the `-walletrejectlongchains` issue, you would replace `9999` with `max_ancestors-1` where `max_ancestors` is defined as in the coin selection code.

fjahr commented:
Rebased and implemented using `max_ancestors - 1` base on feedback.

fjahr commented:
Rebased and gave test a less generic name.

jonatack commented on src/wallet/wallet.h:
```diff
@@ -805,7 +805,7 @@ class CWallet final : public WalletStorage, private interfaces::Chain::Notificat
     void SetUsedDestinationState(WalletBatch& batch, const uint256& hash, unsigned int n, bool used, std::set<CTxDestination>& tx_destinations) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
 
-    std::vector<OutputGroup> GroupOutputs(const std::vector<COutput>& outputs, bool single_coin) const;
+    std::vector<OutputGroup> GroupOutputs(const std::vector<COutput>& outputs, bool single_coin, const size_t max_ancestors = 10000) const;
```
Where does the value of `10000` come from? Could it be documented and maybe hoisted to a static constant?

kallewoof replied:
I think this should be based on the wallet-determined max ancestor count, and should not have a default value here. (i.e. I suggest removing ` = 10000`)

jonatack replied:
Right, `(gArgs.GetArg("-limitancestorcount", DEFAULT_ANCESTOR_LIMIT)` which defaults to 25.

kallewoof replied:
Yep!

fjahr replied:
I removed the default value (10000), thanks!

jonatack commented on test/functional/wallet_avoidreuse.py:
```diff
@@ -288,5 +290,36 @@ def test_getbalances_used(self):
         assert_unspent(self.nodes[1], total_count=2, total_sum=6, reused_count=1, reused_sum=1)
         assert_balances(self.nodes[1], mine={"used": 1, "trusted": 5})
 
+    def test_large_destination_group(self):
```
Suggestion for line 202 if you have to retouch this commit:
```diff
-        self.log.info("Test fund send fund send")
+        self.log.info("Test fund send fund send {}".format(second_addr_type))
```

fjahr replied:
Improved the naming overall in a separate commit. Thanks!

kallewoof commented on test/functional/wallet_avoidreuse.py:
```diff
@@ -288,5 +290,36 @@ def test_getbalances_used(self):
+        should use 10 outputs from the reused address as inputs and not a
+        single 1 BTC input, in order to join several outputs from the reused
+        address.
+        '''
```
I think having a test where the code DOES use the discouraged <10 leftover group would be helpful to ensure that this is correctly chosen when necessary.

fjahr replied:
Added it, thanks.

jonatack commented:
Concept ACK. Can this PR have a `review club` tag to help people know there is a resource?

Reviewers, review club notes and meeting logs of both sessions are here: https://bitcoincore.reviews/17824

Some of the suggestions from the sessions:
- the PR title could use an improvement
- the commit message could be a bit more elaborate
- the changes could be better documented
 
(Also, there seemed to be some agreement in the sessions to consider raising `OUTPUT_GROUP_MAX_ENTRIES`.)

promag commented on src/wallet/wallet.cpp:
```diff
@@ -4023,10 +4023,11 @@ bool CWalletTx::IsImmatureCoinBase() const
+std::vector<OutputGroup> CWallet::GroupOutputs(const std::vector<COutput>& outputs, bool single_coin, const size_t max_ancestors) const {
     std::vector<OutputGroup> groups;
     std::map<CTxDestination, OutputGroup> gmap;
     CTxDestination dst;
```
IMO this should be in the loop.

fjahr replied:
done

fjahr committed:
test: Improve naming and logging of avoid_reuse tests

fjahr commented on src/wallet/wallet.cpp:
```diff
@@ -4040,6 +4041,7 @@ std::vector<OutputGroup> CWallet::GroupOutputs(const std::vector<COutput>& outpu
                 if (gmap[dst].m_outputs.size() >= OUTPUT_GROUP_MAX_ENTRIES) {
                     groups.push_back(gmap[dst]);
                     gmap.erase(dst);
```
done

fjahr commented on src/wallet/wallet.cpp:
```diff
@@ -4040,6 +4041,7 @@ std::vector<OutputGroup> CWallet::GroupOutputs(const std::vector<COutput>& outpu
                 if (gmap[dst].m_outputs.size() >= OUTPUT_GROUP_MAX_ENTRIES) {
```
done

fjahr commented on src/wallet/wallet.cpp:
```diff
@@ -2358,6 +2358,13 @@ bool CWallet::SelectCoins(const std::vector<COutput>& vAvailableCoins, const CAm
             ++it;
     }
 
+    unsigned int limit_ancestor_count;
```
done

fjahr commented on src/wallet/wallet.cpp:
```diff
@@ -2358,6 +2358,13 @@ bool CWallet::SelectCoins(const std::vector<COutput>& vAvailableCoins, const CAm
+    unsigned int limit_ancestor_count;
+    unsigned int limit_descendant_count;
+    chain().getPackageLimits(limit_ancestor_count, limit_descendant_count);
+    size_t max_ancestors = (size_t)std::max<int64_t>(1, limit_ancestor_count);
```
I don't know. I did some digging and the type was introduced [here](https://github.com/bitcoin/bitcoin/pull/12634/files#diff-b2bb174788c7409b671c46ccc86034bdR2588) and prior to that a `uint64` was first introduced [here](https://github.com/bitcoin/bitcoin/pull/9262/files#diff-b2bb174788c7409b671c46ccc86034bdR2020) in this context. I can only speculate that since this is a value from user input it was supposed to be as forgiving as possible if a ridiculously high number is provided, but even that does not make too much sense to me.

kallewoof replied:
This may have been a way to safeguard against a user defined value that was negative, but the switch to `getPackageLimits` removed that safeguard. Not sure if that was wise or not, but that's a separate question.

fjahr commented:
Thanks for all the feedback! I should have addressed all the points here and also what was discussed in the review club. Also I tried to improve the naming of this PR/commit and expanded the commit message.

kallewoof reviewed (approved):
Code review ACK 9dcafb06dfa3f69bc720b5a4f1134fab95f17e40

(Sorry about the horribly named test functions, btw. Thanks for fixing them.)

jonatack reviewed (commented):
ACK 9dcafb06dfa code review and built/tested on the branch and rebased on master. Tried different values in the new tests to verify they hold with different edge cases.

A few comments below but please ignore to not lose review unless you need to retouch this. Thank you for improving the first commit message -- it's very good now; two nits if you have to retouch: s/use/user/ and s/smallet/smaller/. Finally, consider updating the PR description with information from the excellent first commit message.

fjahr commented:
@jonatack Thanks for your comments! For now, I am saving the ACKs and included these cleanups in my followup PR #18418

fjahr commented:
Rebased, no code changes.

jonatack commented:
Re-ACK a2324e4

kallewoof reviewed (commented):
ACK a2324e4d3f47f084b07a364c9a360a0bf31e86a0

achow101 commented:
ACK a2324e4d3f47f084b07a364c9a360a0bf31e86a0

meshcollider reviewed (commented):
Tested ACK a2324e4d3f47f084b07a364c9a360a0bf31e86a0 (verified the new test fails on master without this change)
